ios/sound.mm

```diff
@@ -24,209 +24,323 @@
+ And because Jamulus uses the same buffer to store input and output data (input is sent to server, then output is fetched from server), we actually
+ use the output callback to read inputdata first, process it, and then copy the output fetched from server to ioData, which will then be played.
+ */
+OSStatus CSound::recordingCallback ( void*                       inRefCon,
+                                     AudioUnitRenderActionFlags* ioActionFlags,
+                                     const AudioTimeStamp*       inTimeStamp,
+                                     UInt32                      inBusNumber,
+                                     UInt32                      inNumberFrames,
+                                     AudioBufferList*            ioData )
 {
 
-  // store buffer size
-  //iCoreAudioBufferSizeMono = audioSession.IOBufferDuration;
+    CSound* pSound = static_cast<CSound*> ( inRefCon );
 
-  // init base class
-  //CSoundBase::Init ( iCoreAudioBufferSizeMono );
+    // setting up temp buffer
+    pSound->buffer.mDataByteSize   = pSound->iCoreAudioBufferSizeMono * sizeof ( Float32 ) * pSound->buffer.mNumberChannels;
+    pSound->bufferList.mBuffers[0] = pSound->buffer;
 
-  // set internal buffer size value and calculate stereo buffer size
-  //iCoreAudioBufferSizeStereo = 2 * iCoreAudioBufferSizeMono;
+    // Obtain recorded samples
 
-  // create memory for intermediate audio buffer
-  //vecsTmpAudioSndCrdStereo.Init ( iCoreAudioBufferSizeStereo );
+    // Calling Unit Render to store input data to bufferList
+    AudioUnitRender ( pSound->audioUnit, ioActionFlags, inTimeStamp, 1, inNumberFrames, &pSound->bufferList );
 
-  return iCoreAudioBufferSizeMono;
-}
+    // Now, we have the samples we just read sitting in buffers in bufferList
+    // Process the new data
+    pSound->processBufferList ( &pSound->bufferList, pSound ); // THIS IS WHERE vecsStereo is filled with data from bufferList
 
-void CSound::Start()
-{
-    // ??
+    Float32* pData = (Float32*) ( ioData->mBuffers[0].mData );
 
-    // call base class
-    //CSoundBase::Start();
-}
+    // copy output data
+    for ( int i = 0; i < pSound->iCoreAudioBufferSizeMono; i++ )
+    {
+        pData[2 * i]     = (Float32) pSound->vecsTmpAudioSndCrdStereo[2 * i] / _MAXSHORT;     // left
+        pData[2 * i + 1] = (Float32) pSound->vecsTmpAudioSndCrdStereo[2 * i + 1] / _MAXSHORT; // right
+    }
 
-void CSound::Stop()
-{
-    // ??
-    // call base class
-    //CSoundBase::Stop();
+    return noErr;
 }
 
-void CSound::GetAvailableInOutDevices()
+void CSound::processBufferList ( AudioBufferList* inInputData, CSound* pSound ) // got stereo input data
 {
-  // https://developer.apple.com/documentation/avfoundation/avaudiosession/1616557-availableinputs?language=objc?
+    QMutexLocker locker ( &pSound->MutexAudioProcessCallback );
+    Float32*     pData = static_cast<Float32*> ( inInputData->mBuffers[0].mData );
```

Comment:
Are we processing float audio or integer audio in Jamulus?  If it's still integer on master, can this be integer, too, please.

Reply:
Check this https://github.com/jamulussoftware/jamulus/blob/e0cac75ee0d5ef0a6f85f6b88367abab9b659853/mac/sound.cpp#L929

Reply:
That's not Jamulus sound, that's `mac/sound.cpp`.

Reply:
Yes, I took the code from there.

Reply:
Yes, but it should match the code in the main part of Jamulus to avoid unnecessary data conversions which slow down processing.

Reply:
Still outstanding.

Reply:
Ok I had a look here. Actually iOS AudioBuffer uses float, so I had to work with float. Then when sending/receiving data to/from CSound, conversion to/from integer is done. 
We just can't choose to use integer only.